### PR TITLE
extend draw.ellipse by orientation

### PR DESCRIPTION
**Pull request by Borda**
## Description
extension of an ellipse to have also parameter orientation as well as ellipse_perimeter
reflecting a "issue" #2135

## Checklist
- [x] Clean style in [the spirit of PEP8](https://www.python.org/dev/peps/pep-0008/)
- [x] [Docstrings for all functions](https://github.com/numpy/numpy/blob/master/doc/example.py)
- [x] Gallery example in `./doc/examples` (new features only)
- [x] Unit tests

[For detailed information on these and other aspects see [scikit-image contribution guidelines](http://scikit-image.org/docs/dev/contribute.html)]

## References
[If this is a bug-fix or enhancement, it closes issue # ]

### CODE REVIEW

**Comment by Borda** [comment]
@sciunto how can I find some more details why the checks failed?

**Comment by alexandrejaguar** [comment]
Hi @borda, I treated that issue on #2352. Where do these PRs differ?

**Comment by Borda** [comment]
@alexandrejaguar in you case putting orientation as fifth parameter, may break compatibility for some previous calling and you use orientation in degrees wile other functions like draw.ellipse_perimeter already use radians.

File: skimage/draw/draw.py [hunk 1]
```diff
@@ -30,6 +45,8 @@ def ellipse(r, c, yradius, xradius, shape=None):
         Image shape which is used to determine the maximum extent of output pixel
         coordinates. This is useful for ellipses which exceed the image size.
         By default the full extent of the ellipse are used.
+    orientation : float, optional (default 0.)
+        Set the ellipse orientation (rotation)
```
Comment (JDWarner) [inline skimage/draw/draw.py]: Specify that this is in radians, relative to the column (c) axis.

File: skimage/draw/draw.py [hunk 2]
```diff
@@ -8,18 +8,33 @@
...
 
 
-def ellipse(r, c, yradius, xradius, shape=None):
+def ellipse(r, c, yradius, xradius, shape=None, orientation=0.):
     """Generate coordinates of pixels within ellipse.
 
+    The ellipse equation:
+    ``((x * cos(alpha) + y * sin(alpha)) / xradius) ** 2
+                + ((x * sin(alpha) + y * cos(alpha)) / yradius) ** 2 = 1``
+
```
Comment (JDWarner) [inline skimage/draw/draw.py]: Please place this info in a new `Notes` section at the bottom of the docstring.

File: skimage/draw/draw.py [hunk 3]
```diff
@@ -8,18 +8,33 @@
...
-    distances = ((r_lim - r) / ry) ** 2 + ((c_lim - c) / rx) ** 2
+    sin_alpha, cos_alpha = np.sin(orientation), np.cos(orientation)
+    x, y = (r_lim - r), (c_lim - c)
+    distances = ((x * cos_alpha + y * sin_alpha) / ry) ** 2 \
+                + ((x * sin_alpha + y * cos_alpha) / rx) ** 2
     return np.nonzero(distances < 1)
 
 
-def ellipse(r, c, yradius, xradius, shape=None):
+def ellipse(r, c, yradius, xradius, shape=None, orientation=0.):
```
Comment (JDWarner) [inline skimage/draw/draw.py]: As these are not kwargs, renaming `yradius` -> `rradius` and `xradius` -> `cradius` should be done for consistency.  It won't break anything.  Also update the relevant part of the docstring.

File: skimage/draw/draw.py [hunk 4]
```diff
@@ -8,18 +8,33 @@
                     _bezier_curve)
 
 
-def _ellipse_in_shape(shape, center, radiuses):
-    """Generate coordinates of points within ellipse bounded by shape."""
+def _ellipse_in_shape(shape, center, radiuses, orientation=0.):
+    """ Generate coordinates of points within ellipse bounded by shape.
+    see: http://www.maa.org/external_archive/joma/Volume8/Kalman/General.html
```
Comment (JDWarner) [inline skimage/draw/draw.py]: Put the reference in a `References` section at the bottom of the docstring - look elsewhere in the package for the proper Sphinx formatting for this.

File: skimage/draw/draw.py [hunk 5]
```diff
@@ -8,18 +8,33 @@
...
+
+    :param shape: (int, int) shape of image
+    :param center: (float, float) position of center
+    :param radiuses: (float, float) size of the two half axis
+    :param orientation: float, orientation in radians
+    :return: [int], [int] coordinates of ellipse
```
Comment (JDWarner) [inline skimage/draw/draw.py]: Please use NumPy documentation standards, e.g., 

``` python
"""
Parameters
----------
shape :  iterable of ints
    Shape of the input image.  Must be length 2.
center : iterable of floats
    (row, column) position of center (in what coordinates?  Presumably index coordinates?).
radii : iterable of floats
    Size of two half axes (for row and column, prior to orientation presumably?)
orientation : float, optional
    Rotation of the ellipse defined by the above, in radians, with respect to the column-axis.

Returns
-------
rows : iterable of ints
    Row coordinates representing values within the ellipse.
cols : iterable of ints
    Corresponding column coordinates representing values within the ellipse.
"""
```
or similar.

File: skimage/draw/draw.py (hunk 3 above)
Comment (JDWarner) [inline skimage/draw/draw.py]: Also, from an API standpoint I think I prefer `rotation` to `orientation`.

File: skimage/draw/draw.py [hunk 6]
```diff
@@ -74,7 +106,8 @@ def ellipse(r, c, yradius, xradius, shape=None):
     shifted_center = center - upper_left
     bounding_shape = lower_right - upper_left + 1
 
-    rr, cc = _ellipse_in_shape(bounding_shape, shifted_center, radiuses)
+    rr, cc = _ellipse_in_shape(bounding_shape, shifted_center, radiuses,
```
Comment (JDWarner) [inline skimage/draw/draw.py]: Please use the proper plural `radii` throughout.

File: skimage/draw/draw.py (hunk 3 above)
Comment (soupault) [inline skimage/draw/draw.py]: `yradius` and `xradius` are already renamed in `master` branch.
@Borda please, do a rebase.

**Comment by alexandrejaguar** [comment]
@borda that should be easy to solve, we wouldn't need another PR to solve the same problem. @jdwarner @soupault I'm only saying that because I submitted #2352 before this and no one checked that PR.

File: skimage/draw/draw.py (hunk 3 above)
Comment (Borda) [inline skimage/draw/draw.py]: @soupault I made pull in master (but it says it is up-to-date) and rebase master nothing changed.

Comment (soupault) [inline skimage/draw/draw.py]: @Borda please, double check :). I'm pretty sure the latest signature is different - https://github.com/scikit-image/scikit-image/blob/master/skimage/draw/draw.py#L20.

**Comment by Borda** [comment]
May I ask what is going on with the "ERROR: Unexpected indentation."
the compilation on appveyor is fine but it crashes on travis-ci for the doc string...

**Comment by sciunto** [comment]
Appveyor does build the doc, so it passes. Check your example, the problem is in there.

**Comment by Borda** [comment]
I see, the fail comes at line 38, which was `'.. rubric:: Notes'`, but it is the same as further `'.. rubric:: Examples'` so is there a list of possible rubrics?

File: skimage/draw/draw.py [hunk 7]
```diff
@@ -56,15 +84,44 @@ def ellipse(r, c, r_radius, c_radius, shape=None):
            [0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
            [0, 0, 0, 0, 0, 0, 0, 0, 0, 0]], dtype=uint8)
 
+    >>> img = np.zeros((10, 12), dtype=np.uint8)
```
Comment (sciunto) [inline skimage/draw/draw.py]: I would remove the previous example in favor of your example, which is more general. One is enough I think and it has the advantage to keep the docstring as short as possible.

**Comment by sciunto** [comment]
Just by eye inspection, I do not detect the problem... I need to test your commit. However, I left a little comment. To me, the section's names are fine.

File: skimage/draw/draw.py [hunk 8]
```diff
@@ -56,15 +84,44 @@ def ellipse(r, c, r_radius, c_radius, shape=None):
...
+           [0, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0],
+           [0, 0, 1, 1, 1, 1, 1, 1, 1, 1, 0, 0],
+           [0, 0, 1, 1, 1, 1, 1, 1, 1, 0, 0, 0],
+           [0, 0, 0, 1, 1, 1, 1, 0, 0, 0, 0, 0],
+           [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]], dtype=uint8)
+
+    Notes
+    -----
+    The ellipse equation:
+    ``((x * cos(alpha) + y * sin(alpha)) / x_radius) ** 2
```
Comment (soupault) [inline skimage/draw/draw.py]: I think that the issue is probably here.
Please, check with the different formatting (https://github.com/scikit-image/scikit-image/blob/master/skimage/transform/_geometric.py#L302).

**Comment by Borda** [comment]
@sciunto @soupault  Do you have any idea what is wrong with the doc string?
@JDWarner Do you agree with changes according review?

File: skimage/draw/draw.py [hunk 9]
```diff
@@ -41,30 +69,45 @@ def ellipse(r, c, r_radius, c_radius, shape=None):
...
+           [0, 0, 1, 1, 1, 1, 1, 1, 1, 1, 0, 0],
+           [0, 0, 1, 1, 1, 1, 1, 1, 1, 0, 0, 0],
+           [0, 0, 0, 1, 1, 1, 1, 0, 0, 0, 0, 0],
+           [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]], dtype=uint8)
+
+    Notes
+    -----
+    The ellipse equation::
+
+    ((x * cos(alpha) + y * sin(alpha)) / x_radius) ** 2
```
Comment (soupault) [inline skimage/draw/draw.py]: I think Sphinx expects an indentation here. Try this:
```
    The ellipse equation::

        ((x * cos(alpha) + y * sin(alpha)) / x_radius) ** 2 +
        ((x * sin(alpha) - y * cos(alpha)) / y_radius) ** 2 = 1
```

**Comment by Borda** [comment]
@JDWarner Are you fine with changes according your review? :)

**Comment by JDWarner** [comment]
Looks good now - thanks for your contribution, @Borda, and sorry for the delay!

**Pull request merged**